pkg/util/containers/v2/metrics/provider/collector_cache.go

```diff
@@ -0,0 +1,128 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package provider
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	contStatsCachePrefix    = "cs-"
+	contNetStatsCachePrefix = "cns-"
+	gcInterval              = 30 * time.Second
+)
+
+type cacheEntry struct {
+	value     interface{}
+	err       error
+	timestamp time.Time
+}
+
+// collectorCache is a wrapper handling cache for collectors.
+// this cache is fully synchronized to minimize locking. If a method is called multiple times in parallel for the same cachey key,
+// it may result in multiple calls to the underlying collector
+type collectorCache struct {
+	collector   Collector
+	cache       map[string]cacheEntry
+	cacheLock   sync.RWMutex
+	gcTimestamp time.Time
+}
+
+// NewCollectorCache returns a new cache dedicated to a collector
+func NewCollectorCache(collector Collector) Collector {
+	return &collectorCache{
+		collector: collector,
+		cache:     make(map[string]cacheEntry),
```

Comment:
Why not reuse the existing `cache` in `pkg/util/cache` ?

Reply:
So I wanted to but it does not play very well with the model we use here.
`pkg/util/cache` is based on an expiration date known at `set()` time and there's no way to retrieve the `set()` timestamp.
This cache is based on caller toleration (`cacheValidity`), so we must know the `set()` timestamp.